include/deal.II/non_matching/fe_values.h

```diff
@@ -0,0 +1,511 @@
+   * @f[
+   * N = \{x \in K : \psi(x) < 0 \}, \\
+   * P = \{x \in K : \psi(x) > 0 \}, \\
+   * S = \{x \in K : \psi(x) = 0 \}.
+   * @f]
+   * As in the QuadratureGenerator class, we refer to $N$, $P$ and $S$ as the
+   * inside, outside, and surface region. RegionUpdateFlags is used to describe
+   * how the FEValues objects which are created by NonMatching::FEValues
+   * should be updated.
+   */
+  struct RegionUpdateFlags
+  {
+    /**
+     * Constructor, sets the UpdateFlags for each region to update_default.
+     */
+    RegionUpdateFlags();
+
+    /**
+     * Flags for the region $\{x \in K : \psi(x) < 0 \}$
+     */
+    UpdateFlags inside;
+
+    /**
+     * Flags for the region $\{x \in K : \psi(x) > 0 \}$
+     */
+    UpdateFlags outside;
+
+    /**
+     * Flags for the region $\{x \in K : \psi(x) = 0 \}$
+     */
+    UpdateFlags surface;
+  };
+
+
+  /**
+   * This class intended to facilitate assembling in immersed (in the sense of
+   * cut) finite element methods, when the domain is described by a level set
+   * function, $\psi : \mathbb{R}^{dim} \to \mathbb{R}$. In this type of
+   * methods, we typically need to integrate over 3 different regions of each
+   * cell, $K$:
+   * @f[
+   * N = \{x \in K : \psi(x) < 0 \}, \\
+   * P = \{x \in K : \psi(x) > 0 \}, \\
+   * S = \{x \in K : \psi(x) = 0 \}.
+   * @f]
+   * Thus we need quadrature rules for these 3 regions:
+   * @image html immersed_quadratures.svg
+   * As in the QuadratureGenerator class, we refer to $N$, $P$ and $S$ as the
+   * inside, outside, and surface region. The constructor of this class takes a
+   * discrete level set function (a DoFHandler and a Vector). When the reinit()
+   * function is called the QuadratureGenerator will be called in the background
+   * to create these immersed quadrature rules. This class then creates
+   * dealii::FEValues objects for the inside/outside region and an
+   * FEImmersedSurface object for the surface region. These objects can then be
+   * accessed through one of the functions:
+   * get_inside_fe_values(),
+   * get_outside_fe_values(), and
+   * get_surface_fe_values().
+   * Since a cut between a cell and the domain can be arbitrarily small, it can
+   * happen that the underlying algorithm generates a quadrature rule with 0
+   * points. This can, for example, happen if the cut is of floating point size.
+   * Since the FEValues-like objects aren't allowed to contain 0 points, the
+   * object that get_inside/outside/surface_fe_values() returns is wrapped in a
+   * boost::optional. This requires us to check if the returned FEValues-like
+   * object contains a value before we use it:
+   * @code
+   * NonMatching::FEValues<dim> fe_values(...);
+   *
+   * for (const auto &cell : dof_handler.active_cell_iterators())
+   *  {
+   *    fe_values.reinit(cell);
+   *
+   *    const boost::optional<const FEValues<dim> &> fe_values_inside =
+   *      fe_values.get_inside_fe_values();
```

Comment:
What was the reason again why we don't use `std::shared_ptr` here?

Reply:
1. `optional` communicates more clearly that the returned object may not contain a value.
2. On non-cut cells, we can set the `optional` to the reference we get from `hp::FEValues::get_present_fe_values()`. A `shared_ptr`'s destructor would try to delete that object.

Reply:
Is it possible to use https://github.com/dealii/dealii/blob/master/include/deal.II/base/std_cxx17/optional.h?

Reply:
Unfortunately not, `boost::optional<T>` allows `T` to be a reference, `std::optional<T>` does not, and I want to avoid copying the `FEValues` object.

Reply:
I agree that `std::optional<T &>` would be the right way and that it does not work is a pity. Our policy is normally that we tend to prefer the standardized  functionalities from `std` once they are available (and if they have the same functionalities). In this case, 99% of the functionality is the same (between `std::optional` and `std::boost`) but one feature, which we find useful is missing. Are there any options what to do in such a situation? @bangerth What is your opinion in this regards?

Reply:
Thinking a bit more about this I think I could return
`const std::optional<FEValues<dim>> &`
If I avoid using `hp::FEValues` internally. Would that be acceptable?

Reply:
I guess your idea is to replace the following lines:

https://github.com/dealii/dealii/blob/542b0ef7d8b58225591b6e22ae0251671fe06350/source/non_matching/fe_values.cc#L355-L356

https://github.com/dealii/dealii/blob/542b0ef7d8b58225591b6e22ae0251671fe06350/source/non_matching/fe_values.cc#L377-L378

with the creation of new `FEValues` instances? I don't think this is a good idea, since you make by that the non-cut case slower (creating new `FEValues` instances is not that cheap since all the memory for internal buffers has to allocated).

Reply:
No, instead of `hp::FEValues` as a member variable, I would need to keep the `dealii::FEValues` objects that I need as a member variable:
```c++
std::vector<std::optional<dealii::FEValues<dim>>> fe_values_full_quadrature;
```
Then I can return a reference to one of these `optional`s based on `active_fe_index`.

Reply:
I see. I din't think about this. That is not elegant but definitely an option if we don't find a better solution. Let's wait a couple of days for feedback from others!

Reply:
What about `std::optional<std::reference_wrapper<...>>`?

Reply:
That would work, but then one would have to write `.get()` everywhere when assembling:
```c++
  std::optional<std::reference_wrapper<FEValues<dim>>> fe_values;
  // fe_values->JxW(0); //Doesn't compile
  fe_values->get().JxW(0); // Compiles
```

Reply:
Till now, I think `std::vector<std::optional<dealii::FEValues<dim>>> fe_values_full_quadrature;` would be the best solution. My question would be if the vector is enough? `hp::FEValues` constructs a 3D table for each mapping-, FE-, quadrature-object. Do we need this full flexibility here or is it enough that only on mapping-, FE-, quadrature-object corresponds?

Reply:
I think it seems like the best solution too.  I think the full table is a bit overkill. If someone needs it in the future, the class can be extended later.

Reply:
I also agree here, I would keep it as `vector<optional<FEValues>>`.

Reply:
@simonsticko Could change the code so that it uses `std::vector<std::optional<dealii::FEValues<dim>>>` and add a `#ifndef DOXYGEN` to the problematic function. I think once these two points are done, the PR is ready to go!

Reply:
Okey, but I don't understand how to fix the doxygen problem. It seems to come from `fe/fe_values.cc`.

Reply:
Add a big ifndef around everything.